### PR TITLE
[heft] Clean up the unnecessary re-implementation of the FileSystem API.

### PR DESCRIPTION
**Pull request by iclanton**

### CODE REVIEW

File: apps/heft/src/utilities/fileSystem/TypeScriptCachedFileSystem.ts [hunk 1]
```diff
@@ -0,0 +1,235 @@
...
+  files: string[];
+  directories: string[];
+}
+
+export interface ICreateHardLinkExtendedOptions extends IFileSystemCreateLinkOptions {
+  preserveExisting: boolean;
+}
+
+interface ICacheEntry<TEntry> {
+  entry?: TEntry;
```
Comment (octogonz) [inline apps/heft/src/utilities/fileSystem/TypeScriptCachedFileSystem.ts:28]: Would `TEntry | undefined` be better here? It's not really "optional", it's possibly undefined.

File: apps/heft/src/utilities/fileSystem/TypeScriptCachedFileSystem.ts [hunk 2]
```diff
@@ -0,0 +1,235 @@
...
+ * It uses an in-memory cache to avoid requests against the disk. It assumes that the disk stays
+ * static after construction, except for writes performed through the TypeScriptCachedFileSystem
+ * instance.
+ */
+export class TypeScriptCachedFileSystem {
+  private _statsCache: Map<string, ICacheEntry<FileSystemStats>> = new Map<
+    string,
+    ICacheEntry<FileSystemStats>
+  >();
+  private _readFolderCache: Map<string, ICacheEntry<IReadFolderFilesAndDirectoriesResult>> = new Map<
```
Comment (octogonz) [inline apps/heft/src/utilities/fileSystem/TypeScriptCachedFileSystem.ts:44]: A handy shorthand is:
```ts
private _readFolderCache: Map<string, ICacheEntry<IReadFolderFilesAndDirectoriesResult>> = new Map();
```

Took me about 5 years to figure that out heheh.

File: apps/heft/src/utilities/fileSystem/TypeScriptCachedFileSystem.ts [hunk 3]
```diff
@@ -0,0 +1,235 @@
...
+
+      await this.createHardLinkAsync(options);
+      return true;
+    }
+  };
+
+  private _sortFolderEntries(folderEntries: fs.Dirent[]): IReadFolderFilesAndDirectoriesResult {
+    // TypeScript expects entries sorted ordinally by name
+    // In practice this might not matter
+    folderEntries.sort((a, b) => {
```
Comment (octogonz) [inline apps/heft/src/utilities/fileSystem/TypeScriptCachedFileSystem.ts:183]: Compare with [Sort.compareByValue()](https://github.com/microsoft/rushstack/blob/0f193ac7afcd2527370ebaa342299bb31be54e27/libraries/node-core-library/src/Sort.ts#L34) for some possible edge cases

File: common/changes/@rushstack/heft/ianc-remove-custom-fs_2020-09-18-21-03.json [hunk 4]
```diff
@@ -0,0 +1,11 @@
+{
+  "changes": [
+    {
+      "packageName": "@rushstack/heft",
+      "comment": "",
+      "type": "none"
```
Comment (octogonz) [inline common/changes/@rushstack/heft/ianc-remove-custom-fs_2020-09-18-21-03.json:6]: Why is this `"none"`? Seems like a pretty significant change.

**Review by octogonz — APPROVED** [review]

File: common/changes/@rushstack/heft/ianc-remove-custom-fs_2020-09-18-21-03.json (hunk 4 above)
Comment (iclanton) [reply common/changes/@rushstack/heft/ianc-remove-custom-fs_2020-09-18-21-03.json:6]: It's just a refactor. Nothing should change for any users.

**Pull request merged**